src/dhtproto/node/neo/request/Update.d

```diff
@@ -0,0 +1,312 @@
+        MessageType ret;
+        ed.receive(
+            ( in void[] payload )
+            {
+                Const!(void)[] payload_slice = payload;
+
+                auto message = *ed.message_parser.getValue!(MessageType)(
+                    payload_slice);
+                with ( MessageType ) switch ( message )
+                {
+                    // Client wants to update the record.
+                    case UpdateRecord:
+                        ret = this.update(channel, key, payload_slice);
+                        break;
+
+                    // Client has decided not to update the record.
+                    case LeaveRecord:
+                        ret = MessageType.Ok;
+                        break;
+
+                    // Client has sent the updated record to another node. It
+                    // can be removed here.
+                    case RemoveRecord:
+                        ret = this.remove(channel, key)
+                            ? MessageType.Ok : MessageType.Error;
+                        break;
+
+                    default:
+                        ed.shutdownWithProtocolError("Unexpected message from client");
+                }
+            }
+        );
+
+        return ret;
+    }
+
+    /***************************************************************************
+
+        1. Parses the hash of the original record value and the updated record
+           value from the message payload.
+        2. Gets the record value from storage and hashes it.
+        3. Compares the hash of the original value (provided by the client) with
+           the hash of the value in storage.
+        4a. If the hashes differ, the record has been updated by another client.
+            The request is rejected.
+        4b. If the hashes match, the new value is written to storage.
+
+        Params:
+            channel = channel to update record in
+            key = key of record to update
+            payload = message payload received from client; contains the orignal
+                hash of the record value and the new record value
+
+        Returns:
+            message type code to return to the client
+
+    ***************************************************************************/
+
+    private MessageType update ( cstring channel, hash_t key,
+        Const!(void)[] payload )
+    {
+        // Read original record value hash and updated value from client.
+        auto ed = this.connection.event_dispatcher();
+        hash_t original_hash;
+        Const!(void)[] new_value;
+        ed.message_parser.parseBody(payload, original_hash, new_value);
+
+        // Get the currently stored record value and hash it.
+        bool exists;
+        hash_t stored_hash;
+        auto ok = this.get(channel, key,
+            ( Const!(void)[] value )
+            {
+                exists = true;
+                stored_hash = Fnv1a(value);
+            }
+        );
```

Comment:
I hope there's a guarantee there's no fiber switch between this delegate executing and the next line. Because if there's one, you end up with the race condition within this small window.

Reply:
Hm, it depends on the storage engine's implementation. In the two extant implementations, no fiber switch can occur.

Reply:
Right. Perhaps is enough just to add a comment here.